MetaMask/metamask-extension
# feat: Snap account async transaction redirect

## **Description**

Progresses: https://github.com/MetaMask/accounts-planning/issues/35

This PR implements the `redirectUser` callback that was added in the [eth-snap-keyring](https://github.com/MetaMask/eth-snap-keyring/pull/136). This callbacks allows us to redirect the user to a given url or direct them with custom instructions to sign an async transaction.

[Design](https://www.figma.com/file/K0csZegZ5XQ6TjfKk3n4Z8/Add-dialog-to-inform-about-redirection-to-snap-site-%2335?type=design&node-id=604-2586&mode=design&t=tRpxcs4qeRiRKxPl-0)

## **Manual testing steps**

1. checkout this [branch](https://github.com/MetaMask/snap-simple-keyring/pull/98) in the snap simple keyring repo (this branch is wip but it will work for testing)
2. run `yarn install` and `yarn start`
3. checkout this branch and run `yarn install && yarn start --build-type flask`
4. in your browser load the dist folder into the extensions via `load unpacked`
5. setup a wallet in MM
6. 3. navigate to http://localhost:8000/
7. click `create account`  in the dapp site and approve the creation flow
8. toggle the `Use Synchronous Approval` button at the top of the page so that it is grey. This will turn on async approvals 
9. navigate to https://metamask.github.io/test-dapp/
10. connect your MM to the snap
11. scroll down and click `Personal Sign`
12. a pop up should open in MM, accept that sign
13. a new pop up should open that matches the [above designs](https://www.figma.com/file/K0csZegZ5XQ6TjfKk3n4Z8/Add-dialog-to-inform-about-redirection-to-snap-site-%2335?type=design&node-id=604-2586&mode=design&t=tRpxcs4qeRiRKxPl-0)

## **Screenshots/Recordings**

Blocked URL
<img width="374" alt="Screenshot 2023-10-12 at 11 16 35 AM" src="https://github.com/MetaMask/metamask-extension/assets/22918444/b2835317-5827-4edb-ad4c-b2382ed24d38">



redirect with message and URL
<img width="378" alt="Screenshot 2023-10-12 at 11 14 55 AM" src="https://github.com/MetaMask/metamask-extension/assets/22918444/a08109a3-cffe-40a4-aadf-b8f70928d5d2">



without message
<img width="370" alt="Screenshot 2023-10-12 at 11 18 23 AM" src="https://github.com/MetaMask/metamask-extension/assets/22918444/35b712b8-8116-42ea-b18b-439314cbe63a">



without URL
<img width="392" alt="Screenshot 2023-10-12 at 11 17 33 AM" src="https://github.com/MetaMask/metamask-extension/assets/22918444/debe40ad-8ebf-49d5-94ba-18da04963a62">




### **Before**

N/A

### **After**


https://github.com/MetaMask/metamask-extension/assets/22918444/05213130-ac5d-4d65-ac10-2fe32a96115a



## **Related issues**
https://github.com/MetaMask/accounts-planning/issues/35

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
